src/librustc_lint/unused.rs

```diff
@@ -354,18 +368,30 @@ impl EarlyLintPass for UnusedParens {
                 }
                 let msg = format!("{} argument", call_kind);
                 for arg in args_to_check {
-                    self.check_unused_parens_core(cx, arg, &msg, false);
+                    self.check_unused_parens_expr(cx, arg, &msg, false);
                 }
                 return;
             }
         };
-        self.check_unused_parens_core(cx, &value, msg, struct_lit_needs_parens);
+        self.check_unused_parens_expr(cx, &value, msg, struct_lit_needs_parens);
+    }
+
+    fn check_pat(&mut self, cx: &EarlyContext, p: &ast::Pat) {
+        use ast::PatKind::*;
+        let (value, msg, struct_lit_needs_parens) = match p.node {
+            Ident(.., Some(ref pat)) => (pat, "optional subpattern", false),
+            Ref(ref pat, _) => (pat, "reference pattern", false),
+            Slice(_, Some(ref pat), _) => (pat, "optional position pattern", false),
+            Paren(_) => (p, "pattern", false),
```

Comment:
@nikomatsakis In this arm, `value` is type `&syntax::ast::Pat` and `&syntax::ptr::P<syntax::ast::Pat>` in the rest. I think creating a new `syntax::ptr::P` could solve the issue, but I have not been able to produce a working example.

Reply:
So, the diff in my previous comment [(pushed to my fork)](https://github.com/zackmdavis/rust/commit/b4c948b4b54119f4f2ba307ca7b191a2e96dc610) isn't quite right, because we end up warning twice in the subpattern cases:

```
zmd@ReflectiveCoherence:~/Code/Misc$ rustc +stage1 unused_paren_pat.rs 
warning: unnecessary parentheses around pattern
 --> unused_paren_pat.rs:4:9
  |
4 |         (_) => {},
  |         ^^^ help: remove these parentheses
  |
  = note: #[warn(unused_parens)] on by default

warning: unnecessary parentheses around optional subpattern
 --> unused_paren_pat.rs:8:14
  |
8 |         _b @ (_) => {},
  |              ^^^ help: remove these parentheses

warning: unnecessary parentheses around pattern
 --> unused_paren_pat.rs:8:14
  |
8 |         _b @ (_) => {},
  |              ^^^ help: remove these parentheses

warning: unnecessary parentheses around reference pattern
  --> unused_paren_pat.rs:12:10
   |
12 |         &(_b) => {},
   |          ^^^^ help: remove these parentheses

warning: unnecessary parentheses around pattern
  --> unused_paren_pat.rs:12:10
   |
12 |         &(_b) => {},
   |          ^^^^ help: remove these parentheses
```

but hopefully this helps get you unstuck?

Reply:
Once we get the behavior we want, we'll also want a [UI test](file:///home/zmd/Code/rustc-guide/book/html/tests/adding.html#guide-to-the-ui-tests) in `src/test/ui/lint`.

Reply:
@zackmdavis Thank you for the examples! An explicit coercion with `let (value, msg): (&ast::Pat, _)` ended up being the trick.

Turns out in the end though (and something you pointed out here) is that sub patterns end up being checked, so I did not end up needing to match on any cases besides `Paren`.

I originally opened this to gather some feedback on the issue above; I'll definitely make sure to add some tests!